Standards/scs-0112-v1-volume-type-standard.md

```diff
@@ -0,0 +1,220 @@
+| qos_specs_id       | None                                                                                                                                                         |
++--------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------+
+```
+
+### Replication
+
+Replication states, whether or not there are multiple replicas of a volume. Thus answers the question, whether the data could survive a node outage. Unfortunately there are two ways replication can be achieved:
+
+1. In the configuration of a volume type. It then is visible as extra_spec in the properties of a volume type.
+2. Via the used backend. Ceph for example provides automatic replication, that does not need to be specified in the volume type. This is currently not visible for users.
+
+To fulfill this recommentation for now, the deployer needs to state the replication in the description of a volume type in the following way:
+The description needs to begin with `[replicated]`, after that any further description is allowed.
+
+**TODO:**
+We want to find a way to also use the internal extra_spec for replication, when the replication is automatically done by the backend. If this is not possible, we would like to introduce another property, which has to be set by the admin, when setting the volume type. Only after that we will have the possibility to automatically check for a volume type with replication.
+
+#### OPTIONAL addition: Number of Replicas
+
+Additionally to the fact, that a volume type is replicated, it OPTIONALLY can be stated in the description or name, how many times the provisioned volume will be replicated.
+
+### Example
+
+One volume type that is configured as an encrypted volume type in a ceph backend, with automated replication would fit both recommendations and will be enough to comply to this part of the volume type standard.
+
+It should look like the following part:
+
+```text
++--------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------+
+| Field              | Value                                                                                                                                                        |
++--------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------+
+| access_project_ids | None                                                                                                                                                         |
+| description        | [encrypted][replicated] Content will be replicated three times to ensure consistency and availability for your data. LUKS encryption is used.                |
```

Comment:
Please make the order of the keyword encrypted, replicated, number of replicas, ... in volume's description and name explicitly, similar to [flavor naming standard](https://github.com/SovereignCloudStack/standards/blob/main/Standards/scs-0100-v3-flavor-naming.md#complete-proposal-for-systematic-flavor-naming).

Reply:
1. I doubt that optional parameters like the number of replicas will ever be made a required standard. From my discussion with csps I think, they will stay optional. Therefore I would not like to include them into some keyword standard. But it should be enough if it is stated in the description.
2. That leaves us with only two aspects to standardize: [encrypted] and [replicated]. They both can occur together or alone. So possible beginnings for the description would be:
     - [encrypted]
     - [replicated]
     - [encrypted][replicated]
     - [replicated][encrypted]
3. We only recommend, that both required aspects are present, but not whether they should be combined or can be stand alone. That makes it harder to require a certain sequence.

So if you insist on having a special sequence I would add this, but I don't see any positive side effects from this.

Reply:
The positive side effect of having a dedicated order of keywords is at user side, IMO. As volume type properties are currently defined as string, a cloud service consumer has to parse strings in order to find a volume type with special behavior or properties. Having a standardized order, string parsing will be facilitated.

Reply:
@anjastrunk do you have a particular language in mind that relies on a certain order of arguments to facilitate string parsing?

imho most languages nowadays support some kind of simple pattern matching where the order is not that important?

overall I have no strong opinion if we should enforce the order, but we maybe should in either case explicitly define if this string is ordered or unordered.

Reply:
@anjastrunk please comment

Reply:
The text in the paragraphs above clearly states:

> The description needs to begin with `[encrypted]` ...

> The description needs to begin with `[replicated]` ...

Well, only one of these can be true. This does need some more work!

Reply:
I adjusted the document so that a special order is required now.

Reply:
@fkr You are right, regular expression will not require any order. So I'm fine with extending the standard by adding, that key word are unordered.